subprojects/execution/src/test/groovy/org/gradle/internal/execution/steps/CaptureStateBeforeExecutionStepTest.groovy

```diff
@@ -0,0 +1,149 @@
+        ]
+
+        when:
+        step.execute(context)
+        then:
+        1 * work.visitImplementations(_) >> { UnitOfWork.ImplementationVisitor visitor ->
+            visitor.visitImplementation(implementationSnapshot)
+            additionalImplementations.each {
+                visitor.visitAdditionalImplementation(it)
+            }
+        }
+        interaction { fingerprintInputs() }
+        1 * delegate.execute { BeforeExecutionContext context ->
+            def state = context.beforeExecutionState.get()
+            state.implementation == implementationSnapshot
+            state.additionalImplementations == additionalImplementations
+        }
+        0 * _
+    }
+
+    def "input properties are snapshotted"() {
+        def inputPropertyValue = 'myValue'
+        def valueSnapshot = Mock(ValueSnapshot)
+
+        when:
+        step.execute(context)
+        then:
+        1 * work.visitInputProperties(_) >> { UnitOfWork.InputPropertyVisitor visitor ->
+            visitor.visitInputProperty("inputString", inputPropertyValue)
+        }
+        1 * valueSnapshotter.snapshot(inputPropertyValue) >> valueSnapshot
+        interaction { fingerprintInputs() }
+        1 * delegate.execute { BeforeExecutionContext context ->
+            def state = context.beforeExecutionState.get()
+            state.inputProperties == ImmutableSortedMap.<String, ValueSnapshot>of('inputString', valueSnapshot)
+        }
+        0 * _
+    }
+
+    def "input file properties are fingerprinted"() {
+        def fingerprint = Mock(CurrentFileCollectionFingerprint)
+
+        when:
+        step.execute(context)
+        then:
+        1 * work.visitInputFileProperties(_) >> { UnitOfWork.InputFilePropertyVisitor visitor ->
+            visitor.visitInputFileProperty("inputFile", "ignored", false, { -> fingerprint })
+        }
+        interaction { fingerprintInputs() }
+        1 * delegate.execute { BeforeExecutionContext context ->
+            def state = context.beforeExecutionState.get()
+            state.inputFileProperties == ImmutableSortedMap.<String, CurrentFileCollectionFingerprint>of('inputFile', fingerprint)
+        }
+        0 * _
+    }
+
+    def "output file properties are fingerprinted"() {
+        def outputFileSnapshot = Mock(FileSystemSnapshot)
+
+        when:
+        step.execute(context)
+        then:
+        1 * work.outputFileSnapshotsBeforeExecution >> ImmutableSortedMap.<String, FileSystemSnapshot>of("outputDir", outputFileSnapshot)
+        interaction { fingerprintInputs() }
+        1 * delegate.execute { BeforeExecutionContext context ->
+            def state = context.beforeExecutionState.get()
+            state.outputFileProperties == ImmutableSortedMap.<String, CurrentFileCollectionFingerprint>of('outputDir', AbsolutePathFingerprintingStrategy.IGNORE_MISSING.emptyFingerprint)
+        }
+        0 * _
+    }
+
+    void fingerprintInputs() {
+        _ * context.work >> work
```

Comment:
Finding `context.work >> work` down here was a big confusing for me while reading the tests. Can this be moved to `setup()` instead?

Reply:
Added to the setup method.